### PR TITLE
Port Driver Station Dashboard Articles from Screensteps

### LINKED ISSUE #340: Port LabVIEW dashboard from screensteps
https://wpilib.screenstepslive.com/s/currentCS/m/driver_station/l/144977-frc-driver-station-labview-dashboard

### PR DESCRIPTION
**Pull request by yellowjaguar5**
Fixes #340. 

Note: Images of driver station log errors were transcribed as text in code blocks to make the text searchable.

Articles to be added:

- [x] [FRC Driver Station LabVIEW Dashboard](https://web.archive.org/web/20180623214323/http://wpilib.screenstepslive.com/s/currentCS/m/driver_station/l/144977-frc-driver-station-labview-dashboard)
- [x] [Using the LabVIEW Dashboard with C++\Java Code](https://web.archive.org/web/20180623214323/http://wpilib.screenstepslive.com/s/currentCS/m/driver_station/l/272692-using-the-labview-dashboard-with-c-java-code)
- [x] [Manually Setting the Driver Station to Start SmartDashboard](https://web.archive.org/web/20180623214323/http://wpilib.screenstepslive.com/s/currentCS/m/driver_station/l/290135-manually-setting-the-driver-station-to-start-smartdashboard)
- [x] [FRC Driver Station Errors/Warnings](https://web.archive.org/web/20180623214323/http://wpilib.screenstepslive.com/s/currentCS/m/driver_station/l/144979-frc-driver-station-errors-warnings)

### CODE REVIEW

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+LabVIEW Dashboard
+-----------------
+
+.. image:: images/driver-station-labview-dashboard/labview-dashboard.png
+
+The Dashboard is broken into two main sections. The left pane is for displaying a camera image. The right pane contains:
+
+- Drive tab that contains indicators for joystick and drive motor values (hooked up by default when used with LabVIEW robot code), a gyro indicator, an Autonomous selection text box, a connection indicator and some controls and indicators for the camera,
+- Basic tab that contains some default controls and indicators,
+- Custom tab for customizing the dashboard using LabVIEW
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:15]: Custom and Basic should be reversed for the 2020 dashboard

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+- Test tab for use with Test Mode in the LabVIEW framework
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:16]: Add custom before Commands - placeholder for custom dashboards

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#. Displays a Gyro value
+
+These indicators (other than the Gyro) are hooked up to appropriate values by default when using the LabVIEW framework. For information on using them with C++/Java code see :doc:`using-the-labview-dashboard-with-c++-java-code`.
+
+Autonomous Selection Box and Connection Indicator
+-------------------------------------------------
+
+.. image:: images/driver-station-labview-dashboard/autonomous-selection-box-and-connection-indicator.png
+
+This year, the Dashboard also contains 2 new elements on the Drive tab:
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:55]: Can delete this line, as they are no longer new.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+#. A text box that can be used to select Autonomous modes. Each language's code templates have examples of using this box to select from multiple autonomous programs.
+#. A connection indicator for the connection to the robot.
+
+Basic
+-----
+
+.. image:: images/driver-station-labview-dashboard/basic.png
+
+The Basic tab contains a variety of pre-populated bi-directional controls/indicators which can be used to control the robot or display information from the robot. The SmartDashboard key names associated with each item are labeled next to the indicator with the exception of the Strings which follow the same naming pattern and increment from DB/String 0 to DB/String 4 on the left and DB/String 5 to DB/String 9 on the right. The LabVIEW framework contains an example of reading from the Buttons and Sliders in Teleop. It also contains an example of customizing the labels in Begin. For more detail on using this tab with C++\Java code see :doc:`using-the-labview-dashboard-with-c++-java-code`.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:65]: Update for 2020 dashboard

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+.. warning:: This is not needed for most installations, try using the appropriate :ref:`Dashboard Type setting <docs/software/driverstation/driver-station:Setup Tab>` for your language first.
+
+Set Driver Station to Default
+-----------------------------
+
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/set-ds-to-default.png
+
+Open the Driver Station software, click on the Setup tab and set the Dashboard setting to Default. **Then close the Driver Station!**
+
+Locate SmartDashboard
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst:15]: Update to include Shuffleboard as well as SmartDashboard

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/open-ds-data-storage-file.png
+
+Browse to "C:\Users\Public\Documents\FRC" and double click on "FRC DS Data Storage" to open it.
+
+DashboardCmdLine
+----------------
+
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/dashboard-cmd-line.png
+
+Locate the line beginning with **DasboardCmdLine**. Replace the string after = with **"java -jar "C:\\Users\\USERNAME\\wpilib\\tools\\SmartDashboard.jar""** where Username is the username determined above. Save the file.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst:32]: See new location above

File: source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+.. image:: images/using-the-labview-dashboard-with-c++-java-code/drive-tab.png
+
+Most of the indicators on the Drive tab utilize arrays. The c++ and Java SmartDashboard classes do not currently support sending arrays.
+
+Basic Tab
+---------
+
+.. image:: images/using-the-labview-dashboard-with-c++-java-code/basic-tab.png
+
+The Basic tab uses a number of keys in the a "DB" sub-table to send/recieve Dashboard data. The LED's are output only, the other fields are all bi-directional (send or recieve).
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst:18]:
```suggestion
The Basic tab uses a number of keys in the a "DB" sub-table to send/receive Dashboard data. The LED's are output only, the other fields are all bi-directional (send or receive).
```

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 8]
```diff
@@ -0,0 +1,127 @@
+FRC Driver Station LabVIEW Dashboard
+====================================
+
+The Dashboard application installed and launched by the FRC Driver Station is a LabVIEW program designed to provide teams with basic feedback from their robot, with the ability to expand and customize the information to suit their needs. This Dashboard application uses Network Tables and contains a variety of tools that teams may find useful.
+
+LabVIEW Dashboard
+-----------------
+
+.. image:: images/driver-station-labview-dashboard/labview-dashboard.png
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:9]: Update to picture of 2020 dashboard

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+#. Camera Settings - This control allows you to change the resolution, framerate and compression of the image stream to the dashboard, click the control to pop-up the configuration.
+#. Bandwidth Indicator - Indicates approximate bandwidth usage of the image stream. The indicator will display green for "safe" bandwidth usage, yellow when teams should use caution and red if the stream bandwidth is beyond levels that will work on the competition field.
+#. Framerate - Indicates the approximate received framerate of the image stream.
+
+Drive
+-----
+
+.. image:: images/driver-station-labview-dashboard/drive.png
+
+The center pane contains a section that provides feedback on the joysticks and drive commands when used with the LabVIEW framework and a section that displays the robot IP and reported battery voltage:
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:41]: delete battery voltage for 2020

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+#. Displays X,Y and Throttle information and button values for up to 2 joysticks when using the LabVIEW framework
+#. Displays values being sent to motor controllers when using LabVIEW framework
+#. Displays a brief chart of the reported battery voltage
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:45]: delete battery voltage for 2020

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+- Commands tab for use with the new LabVIEW C&C Framework
+- Checklist tab that can be used to create task lists to complete before and/or between matches
+- Variables tab that displays the raw Network Tables variables in a tree view format
+
+The LabVIEW Dashboard also includes Record/Playback functionality, located in the bottom right. More detail about this feature is included below under `Record/Playback`_.
+
+Camera Image and Controls
+-------------------------
+
+.. image:: images/driver-station-labview-dashboard/camera-image-and-controls.png
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:26]: Update picture for 2020

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+#. Camera Image Display
+#. Mode Selector - This drop-down allows you to select the type of camera display to use. The choices are Camera Off, USB Camera SW (software compression), USB Camera HW (hardware compression) and IP Camera (Axis camera). Note that the IP Camera setting will not work when your PC is connected to the roboRIO over USB.
+#. Camera Settings - This control allows you to change the resolution, framerate and compression of the image stream to the dashboard, click the control to pop-up the configuration.
+#. Bandwidth Indicator - Indicates approximate bandwidth usage of the image stream. The indicator will display green for "safe" bandwidth usage, yellow when teams should use caution and red if the stream bandwidth is beyond levels that will work on the competition field.
+#. Framerate - Indicates the approximate received framerate of the image stream.
+
+Drive
+-----
+
+.. image:: images/driver-station-labview-dashboard/drive.png
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:39]: Update picture for 2020

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+---------
+
+.. image:: images/driver-station-labview-dashboard/variables.png
+
+The Variables tab of the left pane shows all Network Tables variables in a tree display. The Variable Name (Key), Value and data type are shown for each variable. Information about the Network Tables bandwidth usage is also displayed in this tab. Entries will be shown with black diamonds if they are not currently synced with the robot.
+
+Record/Playback
+---------------
+
+.. image:: images/driver-station-labview-dashboard/record-playback.png
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:105]: Update picture for 2020

File: source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+==============================================
+
+The default LabVIEW Dashboard utilizes Network Tables to pass values and is therefore compatible with C++ and Java robot programs. This article covers the keys and value ranges to use to work with the Dashboard.
+
+Drive Tab
+---------
+
+.. image:: images/using-the-labview-dashboard-with-c++-java-code/drive-tab.png
+
+Most of the indicators on the Drive tab utilize arrays. The c++ and Java SmartDashboard classes do not currently support sending arrays.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst:11]: The C++ and Java classes now support arrays, so this comment can be deleted

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 11 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:26]: When updating the picture I noticed that there is no bandwidth indicator or framerate on the dashboard when no robot is connected. Is this correct and if so, should the notes about the bandwidth indicator and framerate be removed? Also, does the image below work?
image

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 2 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:16]: Does this sequence make sense?

```
The Dashboard is broken into two main sections. The left pane is for displaying a camera image. The right pane contains:

- Drive tab that contains indicators for joystick and drive motor values (hooked up by default when used with LabVIEW robot code), a gyro indicator, an Autonomous selection text box, a connection indicator and some controls and indicators for the camera
- Basic tab that contains some default controls and indicators
- Test tab for use with Test Mode in the LabVIEW framework
- Custom tab for customizing the dashboard using LabVIEW
- Commands tab for use with the new LabVIEW C&C Framework
- Checklist tab that can be used to create task lists to complete before and/or between matches
- Variables tab that displays the raw Network Tables variables in a tree view format

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 3 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst]: I reworked this section to combine it with the section above so all information about the `Drive` tab stays together. Is that okay?

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 4 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst]: My version of the 2020 dashboard looks identical to the description. Is there a change I should be noting?

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst (hunk 5 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst]: Fixed in 1db7c47.

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst (hunk 6 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-smartdashboard.rst]: Fixed in 1db7c47.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 8 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:9]: Updated to:
image

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 9 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst]: I removed the note about battery voltage and **robot IP** in 5a7d41c. Is that correct?

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 10 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst]: Removed in 5a7d41c.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 12 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:39]: Updated to: 
image

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 13 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:96]: Updated to: 
image

File: source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst (hunk 14 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst]: Removed in 6e8cea0.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 4 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst]: I had misread the description

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 2 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:15]: Missing Camera tab between drive and basic.
The order should match the dashboard
drive, basic, camera, custom, test, commands, checklist, variables

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 11 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:26]: When there is a camera connected, the bandwidth indicator does appear, so the description should remain. 
image

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+-----------------------------
+
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/set-ds-to-default.png
+
+Open the Driver Station software, click on the Setup tab and set the Dashboard setting to Default. **Then close the Driver Station!**
+
+Locate Dashboard JAR file
+-------------------------
+
+Find the location of the dashboard jar file (SmartDashboard or Shuffleboard). It should be installed to C:\Users\USERNAME\wpilib\tools where USERNAME is the username of the user logged into the computer.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:18]: Need to escape the path

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+Open DS Data Storage file
+-------------------------
+
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/open-ds-data-storage-file.png
+
+Browse to "C:\Users\Public\Documents\FRC" and double click on "FRC DS Data Storage" to open it.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:25]: Need to escape path

File: source/docs/software/driverstation/troubleshooting-dashboard-connectivity.rst [hunk 17]
```diff
@@ -52,4 +52,6 @@ If the Dashboard does not connect to the Robot (after the Driver Station has con
 
 2. If that doesn't work, restart the Robot Code using the Restart Robot Code button on the Diagnostics tab of the Driver Station
 
-3. If it still doesn't connect verify that the Team Number is set properly in the Dashboard and that your Robot Code writes a SmartDashboard value during initialization or disabled
+3. If it still doesn't connect, verify that the Team Number is set properly in the Dashboard and that your Robot Code writes a SmartDashboard value during initialization or disabled
+
+4. Finally, if it still doesn't connect, follow the steps in the :doc:`manually-setting-the-driver-station-to-start-dashboard` article
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/troubleshooting-dashboard-connectivity.rst:57]: Delete this step. It doesn't help with connections

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst (hunk 16 above)
Comment (Daltz333) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:25]: Or better, replace it with an inline formatting block ``C:\Users\Public\Documents\FRC``

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst (hunk 15 above)
Comment (Daltz333) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:18]: Or better, replace it with an inline formatting block ``C:\Users\Public\Documents\FRC``

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+----------------
+
+.. image:: images/manually-setting-the-driver-station-to-start-smartdashboard/dashboard-cmd-line.png
+
+Locate the line beginning with **DasboardCmdLine**.
+
+For SmartDashboard
+^^^^^^^^^^^^^^^^^^
+
+Replace the string after = with **"java -jar "C:\\Users\\USERNAME\\wpilib\\tools\\SmartDashboard.jar""** where Username is the username determined above. Save the file.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:37]: Should add a note somewhere that java 11 is required for SmartDashboard and Shuffleboard

Comment (Daltz333) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:37]: It'll use the built in wpilib java 11

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 11 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:26]: Thanks for the clarification. Could I use this screenshot (as I don't have a roboRIO nearby) and could you send another with a camera connected to the camera tab (on the right pane)?

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst (hunk 18 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: If wpilib was installed with the installer, this article isn't necessary as the tools would launch automatically, This is specifically for someone who's trying to set up a driver station without installing all of wpilib.

Also, as wpilib doesn't put the wpilib java on the path, just running java -jar wouldn't run the wpilib jdk, you would have to point to the path of the wpilib jdk.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 11 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:27]: Here's one with the camera on the right

image

Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:27]: Thanks. I find it odd that there is not a framerate or bandwidth indicator in the bottom bar below the camera tab.

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 2 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:17]: Fixed in 36286f9.

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst (hunk 18 above)
Comment (yellowjaguar5) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: I'll add a note that to use a dashboard without WPILib, Java 11 is required. Does that sound good?

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst (hunk 11 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:27]: The bandwidth indicator is a combined bandwidth. The framerate indicator is by the bandwidth indicator. 
image

Comment (yellowjaguar5) [reply source/docs/software/driverstation/driver-station-labview-dashboard.rst:27]: That makes a lot of sense. I'll include a note in the camera tab section. Thanks!

File: source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst [hunk 19]
```diff
@@ -0,0 +1,110 @@
+Using the LabVIEW Dashboard with C++\Java Code
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/using-the-labview-dashboard-with-c++-java-code.rst:1]:
```suggestion
Using the LabVIEW Dashboard with C++/Java Code
```

File: source/docs/software/driverstation/driver-station-labview-dashboard.rst [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+.. image:: images/driver-station-labview-dashboard/custom.png
+
+The Custom tab allows you to add additional controls/indicators to the dashboard using LabVIEW without removing any existing functionality. To customize this tab you will need to create a Dashboard project in LabVIEW.
+
+Test
+----
+
+.. image:: images/driver-station-labview-dashboard/test.png
+
+The Test tab is for use with Test mode for teams using LabVIEW (Java and C++ teams should use the Java SmartDashboard when using Test Mode). For many items in the libraries, Input/Output info will be populated here automatically. All items which have ** next to them are outputs that can be controlled by the dashboard. To control an output, click on it to select it, drag the slider to set the value then press and hold the green button to enable the output. As soon as the green button is released, the output will be disabled. This tab can also be used to run and monitor tests on the robot. An example test is provided in the LabVIEW framework. Selecting this test from the dropdown box will show the status of the test in place of the slider and enable controls.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/driver-station-labview-dashboard.rst:90]:
```suggestion
The Test tab is for use with Test mode for teams using LabVIEW (Java and C++ teams should use SmartDashboard or Shuffleboard when using Test Mode). For many items in the libraries, Input/Output info will be populated here automatically. All items which have ** next to them are outputs that can be controlled by the dashboard. To control an output, click on it to select it, drag the slider to set the value then press and hold the green button to enable the output. As soon as the green button is released, the output will be disabled. This tab can also be used to run and monitor tests on the robot. An example test is provided in the LabVIEW framework. Selecting this test from the dropdown box will show the status of the test in place of the slider and enable controls.
```

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst [hunk 21]
```diff
@@ -0,0 +1,47 @@
+Manually Setting the Driver Station to Start Dashboard
+======================================================
+
+.. note:: If WPILib is not installed to the default location (such as when files are copied to a PC manually), the dashboard of choice may not launch properly. To have the DS start a dashboard when it starts up, you have to manually modify the settings for the default dashboard.
+
+.. warning:: This is not needed for most installations, try using the appropriate :ref:`Dashboard Type setting <docs/software/driverstation/driver-station:Setup Tab>` for your language first. Also, to use a dashboard without installing WPILib, Java 11 must be installed first.
```
Comment (sciencewhiz) [inline source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst:6]:
```suggestion
.. warning:: This is not needed for most installations, try using the appropriate :ref:`Dashboard Type setting <docs/software/driverstation/driver-station:Setup Tab>` for your language first. Also to use SmartDashboard or Shuffleboard, Java 11 must be installed first.
```

File: source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst (hunk 18 above)
Comment (sciencewhiz) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: Even if WPILib is installed, just calling out java -jre would run the system java, not the wpilib java installation, so it's still necessary to have java 11 installed as the system java.

Comment (Daltz333) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: You could add a note to run ``frcvars2020``. I would encourage this instead.

Comment (sciencewhiz) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: As the wpilib java doesn't get security updates (except yearly), I don't recommend setting it as the system java. Thus you'd have to make a batch file that runs frcvars and then the dashboard. But that still assumes that wpilib is installed, in which case this article isn't necessary. I think the current wording (that incorporated my suggested change) is good.

Comment (Daltz333) [reply source/docs/software/driverstation/manually-setting-the-driver-station-to-start-dashboard.rst]: I don't like the idea of supporting shuffleboard/smartdashboard outside of the installer. It only leads to problems.

**Comment by yellowjaguar5** [comment]
@sciencewhiz Could you review the latest changes on this PR?

**Pull request merged**